pkg/infradb/subscriberframework/actionbus/actionbus.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2024 Ericsson AB
+
+// Package actionbus holds implementation for subscribing and receiving actions
+package actionbus
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+// ABus holds the ActionBus object
+var ABus = NewActionBus()
+
+// ActionBus holds the action bus info
+type ActionBus struct {
+	subscribers    map[string][]*Subscriber
+	actionHandlers map[string]ActionHandler
+	subscriberL    sync.RWMutex
+	mutex          sync.RWMutex
+}
+
+// Subscriber holds the info for each subscriber
+type Subscriber struct {
+	Name string
+	Ch   chan interface{}
+	Quit chan bool
+}
+
+// ActionHandler handles the action requests that arrive
+type ActionHandler interface {
+	HandleAction(string, *ActionData)
+}
+
+// ActionData holds the data for each action
+type ActionData struct {
+	ErrCh chan error
+}
+
+// StartSubscriber will be called by the modules to initialize and start listening for actions
+func (a *ActionBus) StartSubscriber(moduleName, actionType string, actionHandler ActionHandler) {
+	if !a.subscriberExist(actionType, moduleName) {
+		subscriber := a.Subscribe(moduleName, actionType, actionHandler)
+
+		go func() {
+			for {
+				select {
+				case action := <-subscriber.Ch:
+					log.Printf("\nSubscriber %s for %s received \n", moduleName, actionType)
+
+					handlerKey := moduleName + "." + actionType
+					if handler, ok := a.actionHandlers[handlerKey]; ok {
+						if actionData, ok := action.(*ActionData); ok {
+							handler.HandleAction(actionType, actionData)
+						} else {
+							log.Println("error: unexpected action type")
+						}
+					} else {
+						log.Println("error: no action handler found")
+					}
+				case <-subscriber.Quit:
+					close(subscriber.Ch)
+					return
+				}
+			}
+		}()
+	}
+}
+
+// NewActionBus initializes an ActionBus object
+func NewActionBus() *ActionBus {
+	return &ActionBus{
+		subscribers:    make(map[string][]*Subscriber),
+		actionHandlers: make(map[string]ActionHandler),
+	}
+}
+
+// NewActionData initializes an ActionData object
+func NewActionData() *ActionData {
+	return &ActionData{
+		ErrCh: make(chan error),
+	}
+}
+
+// Subscribe api provides registration of a subscriber to the given action
+func (a *ActionBus) Subscribe(moduleName, actionType string, actionHandler ActionHandler) *Subscriber {
+	a.subscriberL.Lock()
+	defer a.subscriberL.Unlock()
+
+	subscriber := &Subscriber{
+		Name: moduleName,
+		Ch:   make(chan interface{}),
+		Quit: make(chan bool),
+	}
+
+	a.subscribers[actionType] = append(a.subscribers[actionType], subscriber)
+	a.actionHandlers[moduleName+"."+actionType] = actionHandler
+
+	log.Printf("Subscriber %s registered for action %s\n", moduleName, actionType)
+	return subscriber
+}
+
+// GetSubscribers api is used to fetch the list of subscribers registered with given actionType
+func (a *ActionBus) GetSubscribers(actionType string) []*Subscriber {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+
+	return a.subscribers[actionType]
```

Comment:
if we expose a slice of pointers to another thread here, how can we be sure that no one would change that slice from another thread? not change elements from another thread?

Reply:
Sorry I am not sure that I get this one. Can you please elaborate ? For which reason someone to change the slice of pointers ?

Reply:
here we return `a.subscribers[actionType]` from this function. It is a slice. Slices are implemented in such a way that they copy just a pointer to actual data. It means that whoever called that function can change subscribers stored in `a.subscribers` (if they assumed that they keep a deep copy of it)

Reply:
Let me think about this

Reply:
The whole implementation is based on pointers so we can have better utilization of memory and alo there are channels for each subscriber that they used for communication between the different threads. So I do not think is that easy to just return a copy of Subscribers for instance and not the pointers of those subscribers as then the system will not fucntion due to the channels that are bound to those subscribers objects. I hear your concern but I honestly cannot see any case where some thread will change a subscriber. Even if a thread can change a subscriber I do not see a real risk that this will happen. Yes is a valid concern but I see low possibility of happening. If we return a copy of the Subscribers and not pointers then the whole implementation I believe needs to change in many places and maybe we need to rethink the whole thing which is not vital right now I believe. Also the code is functioning as one we do not have plugins that come and go during runtime so if the code passes the integration tests and the tests in general then I do not see how some thread that have been introduced through some PR will change the subscribers and the integration test gate will give a green light.